[IN REVIEW] Broadcaster / Listener pattern to suceed Observers / Observerables

henrij22 opened the pull request:
Updated from https://github.com/ikarus-project/ikarus/pull/346
Original idea: https://godbolt.org/z/b9cM5e714

This is now to be merged into **main**

There is a test [testelementsaslisteners.cpp](https://github.com/ikarus-project/ikarus/blob/feature/broadcaster/tests/src/testelementsaslisteners.cpp) that details the implementation of listeners for skills

### ToDo: 

- [x] Implement a nice `unscribe` functionality (if possible) 
- [x] Fix type of signature the solvers are emitting to the elements (`updateState`)

### Done: 

- [x] Implement Broadcasters and Listeners
- [x] Change NRSolvers to Broadcasters
- [x] Change the existing loggers to listeners
- [x] Functionality to have the skills tell the mixin what functions they want to listen to 
- [x] Register the fes to Control routines and NR Solvers in the constructor of control routines

### Resolved issues (for now)

- [x] Investigate the memory bug (i fixed it by moving the SubsidaryArgs to a member variable of PF, but maybe there is an issue somewhere in the broadcasters?) 
- [x] Investigate the compiler divergence in requires clause https://github.com/ikarus-project/ikarus/blob/21f39f81a140bf3bf3ed71483c41bad453e9b468/ikarus/finiteelements/mixin.hh#L253 see also comment https://github.com/ikarus-project/ikarus/pull/349#discussion_r1924763239

henrij22 commented on ikarus/finiteelements/mixin.hh:
```diff
@@ -244,6 +247,26 @@ public:
+private:
+  template <typename Sk, typename BC, typename MT>
+  auto invokeSubscribeTo(BC& bc) {
+#if defined(__clang__)
```
For Clang: we need the `this->` otherwise the code in the if clause will never be called. 
For Gcc: with the `this->` it throws a compiler error in certain cases

rath3t replied:
This is not nice but thats how it is. Can you write the comment you wrote here also in the code. Also with the current compilers version, because we can check that in the future maybe.

rath3t commented on ikarus/assembler/dirichletbcenforcement.hh:
```diff
@@ -1,8 +1,9 @@
-#include <ikarus/utils/makeenum.hh>
 
 #pragma once
+#include <ikarus/utils/makeenum.hh>
```
Ah nice catch! Thanks

rath3t commented on ikarus/controlroutines/controlinfos.hh:
```diff
@@ -25,4 +27,10 @@ struct ControlInformation
   int totalIterations{0}; ///< Total number of iterations performed.
```
What happenedn here strange diff? Maybe lineendings ?

henrij22 replied:
I added this NoAssembler struct, but that can actually go somewhere else

rath3t commented on ikarus/controlroutines/loadcontrol.hh:
```diff
@@ -51,6 +56,13 @@ public:
         }, "The last parameter (load factor) must be assignable and incrementable with a double!");
+
+    // register FEs to listen to NR messages
+    if constexpr (not std::is_same_v<Assembler, Impl::NoAssembler>)
```
To keep the control routines clean this might be more suitable in a factory function?

rath3t replied:
Then all controlroutine factories can call the same function which contains the fe subsribe mechanism, which would reduce code duplciation

henrij22 replied:
Ah yes we can do this, I think we discussed this already

rath3t commented on ikarus/controlroutines/pathfollowing.inl:
```diff
@@ -26,47 +25,48 @@ requires(Impl::checkPathFollowingTemplates<NLS, PF, ASS>())
+    subsidiaryArgs_.currentStep = ls;
 
-    adaptiveStepSizing_(solverInfo, subsidiaryArgs, nonOp);
+    adaptiveStepSizing_(solverInfo, subsidiaryArgs_, nonOp);
```
Was the fix of the memory issue related to making subsidiary args a member variable?

henrij22 replied:
Yup, that helped keeping the memory away from there, but maybe there is something fishy going on at the broadcasters anyway?

rath3t replied:
Hm I don't know. BUt we will see. If it works for now.-

rath3t commented on ikarus/controlroutines/pathfollowingfunctions.hh:
```diff
@@ -44,6 +44,16 @@ struct SubsidiaryArgs
   int currentStep;              ///< The current step index in the control routine.
+
+  void reset() {
+    stepSize = 0.0;
```
What should this `reset` actually do? Reset means for me that the vectors should also be resized to zero, so resetting to default state.
If this should behave like it is implemented now and the size of the vectors is already known, I suggest renaming it to setZero() instead of reset().

rath3t replied:
You can als pass than the first parameter of the nonlinOP

rath3t commented on ikarus/controlroutines/controlroutinefactory.hh:
```diff
@@ -0,0 +1,62 @@
+
+    for (auto& fe : assembler->finiteElements()) {
+      fe.template subscribeTo<NonLinearSolverMessages>(nonlinearSolver);
+      fe.template subscribeTo<ControlMessages>(cr);
```
Thats nice and seperated

rath3t commented on ikarus/controlroutines/controlinfos.hh:
```diff
@@ -9,6 +9,8 @@
 
 #pragma once
 
+#include <vector>
```
This include should not ne needed here

henrij22 replied:
mhm with iwyu in mind, I think it should be here ... also clangd complaines otherwise, but i can of course remove it

rath3t replied:
Ah you are right, I overlooked the std::vector inside the controlinformation

rath3t commented on ikarus/controlroutines/pathfollowing.hh:
```diff
@@ -45,6 +46,59 @@ namespace Impl {
+
+template <typename PF>
+PathFollowingConfig(int, double, PF) -> PathFollowingConfig<PF>;
+
```
Maybe these can also be added to the nonlinear solver configs because there with the designated initializers it didnt work for me back than. There was also a problem with clang or gcc.

henrij22 replied:
i am not yet sure what to do with the settings, but i could simplify it a bit for example here 
https://github.com/ikarus-project/ikarus/blob/d32b00b8b69ffbefa9b73dc0eb05c68c190d6eda/ikarus/finiteelements/mechanics/materials/vanishingstress.hh#L198
or 
https://github.com/ikarus-project/ikarus/blob/d32b00b8b69ffbefa9b73dc0eb05c68c190d6eda/tests/src/testtruss.cpp#L117

The first argument is for now always the settings, but they can be just `{}` or no argument, then it should default to everything

rath3t replied:
That seems fine for now

rath3t commented on ikarus/controlroutines/pathfollowing.inl:
```diff
@@ -26,47 +25,44 @@ requires(Impl::checkPathFollowingTemplates<NLS, PF, ASS>())
   ControlInformation info;
   auto& nonOp = nonLinearSolver_->nonLinearOperator();
-  this->notify(ControlMessages::CONTROL_STARTED, pathFollowingType_.name());
+  this->notifyListeners(ControlMessages::CONTROL_STARTED, pathFollowingType_.name());
```
Should we keep `notify` or `notifyListeners`? I would like it shorted because otherwise it pollutes the classes even more but I also like the more destriptive nature of  `notifyListeners`.

henrij22 replied:
Yeah i think `notfiy` is just fine .. i had it like that because i had the two systems in place simultaniously

rath3t commented on ikarus/controlroutines/pathfollowingfunctions.hh:
```diff
@@ -44,6 +44,18 @@ struct SubsidiaryArgs
   double dfdDlambda;            ///< The derivative of the subsidiary function with respect to Dlambda.
   int currentStep;              ///< The current step index in the control routine.
+
+  void setZero(const auto& firstParameter) {
```
Can you add the type here? Either `Eigen::VectorX<double>` or writing it using the Eigen::MatrixBase<Derived stuff> or we can also use a more modern solution of using cocnepts:

```cpp
template <typename T>
concept EigenType = std::is_base_of_v<Eigen::MatrixBase<std::decay_t<T>>, std::decay_t<T>>;
```

Then we can skip the `MatrixBase` shenegans everywhere

henrij22 replied:
Okay nice, in this PR? :D

henrij22 replied:
You can archieve this then in the function signature with 
```cpp
fun(const EigenType auto&) 
```
or 
```cpp
template <EigenType Matrix>
fun(const Matrix&) 
```

rath3t replied:
The upper one is fine. Thanks!

rath3t commented on ikarus/utils/broadcaster/broadcaster.hh:
```diff
@@ -0,0 +1,100 @@
+ * \tparam Args the arguments of the signature the Broadcaster can emit.
+ */
+template <typename... Args>
+class Broadcaster<void(Args...)>
```
Do we also need things returning not `void` later? Can you think of a use case?

henrij22 replied:
I thought about this as well but I couldn't find a use case. But maybe a return type could be for example a `bool` for a abort criterion? :D

rath3t replied:
Using listeners would be also an interesting options for #341

rath3t commented on ikarus/utils/listener/listener.hh:
```diff
@@ -0,0 +1,74 @@
+   */
+  template <typename Broadcaster, typename Signature, typename F>
+  void subscribe(Broadcaster& broadcaster, F&& f) {
+    if constexpr (requires { broadcaster.operator->(); })
```
Would it be better to only allow one ( references or pointer) to the broadcaster here? Then it is more obvious what is expcected here

henrij22 replied:
The thing is, that our nonlinearsolvers are most often a shared_ptr and the load controlS are values. 
This makes it easier for the user side. So he doens't need to care. 
I think its fine to have here the additional line

rath3t commented on ikarus/utils/traits.hh:
```diff
@@ -337,8 +337,9 @@ struct FunctionTraits;
-  using return_type = R;
-  using ArgsTuple   = std::tuple<Args...>;
+  using return_type   = R;
+  using ArgsTuple     = std::tuple<Args...>;
```
I found https://gitlab.dune-project.org/staging/dune-functions/-/blame/master/dune/functions/common/signature.hh?ref_type=heads#L89 , which seems to be what we want. Can you try to throw the whole FunctionTraits out?

henrij22 replied:
I unfortunatly get 
```
no type named 'RawSignature' in 'Dune::Functions::SignatureTraits<(lambda at /workspaces/ikarus/tests/src/testelementsaslisteners.cpp:63:30), true>'
```

rath3t replied:
Hmm for me it works also in DifferentiableFunction. Can you elaborate?

henrij22 replied:
Mhm that was back in January when i tried this. But I think I had the same problem with lambdas that we had before

rath3t replied:
I think this only fails when the signature of the lambda is not already clear. Maybe yxou explicitly hgave to specify --> void in the lambdas? But Im nopt sure though

henrij22 replied:
Should I try again? I am not exactly sure what you mean with specifying void. Or should we leave it as such. I think its not too bad, because we have the FunctionTraits already?

rath3t replied:
Yes it is fine. I meant that you explicitly have to specify the retuirn type of the lambda to get it working. But we can keep it as such

rath3t commented on tests/src/testadaptivestepsizing.cpp:
```diff
@@ -154,33 +154,41 @@ auto KLShellAndAdaptiveStepSizing(const PathFollowingType& pft, const std::vecto
-      [&]() { nonLinearSolverObserver->update(Ikarus::NonLinearSolverMessages::BEGIN); },
-      "nonLinearSolverObserver should have failed for the BEGIN message");
+  // TODO We lost this functionality, we could of course re-enable it somewhere
+  // t.checkThrow<Dune::InvalidStateException>(
```
Thats no problem or is it? just remove it

henrij22 replied:
I don't think so

rath3t commented on tests/src/testelementsaslisteners.cpp:
```diff
@@ -0,0 +1,242 @@
+      return std::make_tuple([&](NonLinearSolverMessages message, Eigen::VectorXd& vec,
+                                 const std::remove_reference_t<typename Traits::template VectorType<>>& dx) {
+        auto req = Requirement();
+        req.insertGlobalSolution(vec);
```
To dummy create here requirements seems not so nice, this would also be the case for EAS? If the current load factor is needed, this does not work. Maybe the finall updateState signature has to be different.
But for this dummy test you can change the updateState signature below to not have this indirection.

henrij22 replied:
Exactly, at the moment we need it here https://github.com/ikarus-project/ikarus/blob/92e25d084298ba8d87ddf308f10c6f951db48b0b/ikarus/finiteelements/mechanics/enhancedassumedstrains.hh#L348
I am not sure exactly what the `dx` could be, but maybe we can pass here the `d`. Maybe @tarun-mitruka can we find an implementatino without the need of the ferequirements?

Yup i will simplify the test

tarun-mitruka replied:
Isn't the `dx` the optional argument for autodiff? This cannot be changed or?

henrij22 replied:
So I think we can pass in the parmeters from NLO here, displacment vector and load parameter, then we can construct a full FERequirements from that. But we need to construct a FERequirement somewhere I think, and the NonlinearSolver doens't know about the requirements i think (?) to this is I think the best idea

rath3t replied:
Sine `dx` is optional it does nt have to be provided.

rath3t reviewed (commented):
This looks really nice and almost done! Thanks!

rath3t commented on ikarus/finiteelements/mechanics/materials/vanishingstress.hh:
```diff
@@ -199,11 +199,7 @@ private:
-        .linearSolver   = linearSolver,
-        .updateFunction = updateFunction
-    };
+    NewtonRaphsonConfig nrs({.tol = tol_, .maxIter = 100, .minIter = minIter}, linearSolver, updateFunction);
```
Ok can you move the CTAD comment from here to the deduction guides inside the newton raphson files to explain why these deduction guides are needed.

rath3t commented on ikarus/solver/nonlinearsolver/newtonraphson.hh:
```diff
@@ -37,7 +38,7 @@ struct NewtonRaphsonConfig
   using LinearSolver   = LS;
   using UpdateFunction = UF;
-  NRSettings parameters;
+  NRSettings parameters{};
```
Can you also default init the linearSovler and updateFunction?

rath3t commented on tests/src/testadaptivestepsizing.cpp:
```diff
@@ -154,33 +154,32 @@ auto KLShellAndAdaptiveStepSizing(const PathFollowingType& pft, const std::vecto
+  vtkWriter.unSubscribeAll();
+  pathFollowingObserver.subscribeTo(crWSS);
+
+  // crWSS.subscribe(ControlMessages::SOLUTION_CHANGED, vtkWriter);
```
Removing or keeping the code? Commiting the comment we should avoid.

rath3t commented on ikarus/utils/listener/listener.hh:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  /**
+   * \brief Unsubscribe from all listeners. At the moment unsubscribing can't be done more granularly.
```
I think you can just return the listener from the subscribe function (just return `t.back()`). Then you can use `unSubscribe(Token)`

henrij22 replied:
For now, I always had the `subscribeTo` method returning a this pointer of the listener itself. That made chaining easier ( I think simliar to FERequirements)

rath3t replied:
Yes but then you can not use that pointer to unsubscribe or do you?

So basically you can write a unSubscribe Function, which accepts just a broadcaster

rath3t replied:
So one way of unsubscribing would be : https://godbolt.org/z/nKojvea11

henrij22 replied:
I think that will work nicely for control loggers and such, but I am afraid not really well for elements:
```cpp
  template <typename MT, typename BC>
  auto subscribeTo(BC& bc) {
    (invokeSubscribeTo<Skills<PreFE, typename PreFE::template FE<Skills...>>, BC, MT>(bc), ...);
    return *this;
```

Every skill can register its own tokens, therfore we have a huge list of tokens which makes it again not really granular. Maybe we can have a list of tokens on skill level, than can be deregisterd, but it is needed?

henrij22 replied:
And also here fore control logger 
```cpp
template <typename BC>
  ControlLogger& subscribeTo(BC& bc) {
    this->subscribe(bc, [&](ControlMessages message) { this->updateImpl(message); });
    this->subscribe(bc, [&](ControlMessages message, const std::string& val) { this->updateImpl(message, val); });
    this->subscribe(
        bc, [&](ControlMessages message, int val1, const std::string& val2) { this->updateImpl(message, val1, val2); });
    this->subscribe(bc, [&](ControlMessages message, int val1, double val2) { this->updateImpl(message, val1, val2); });
    return *this;
  }
  ```
  
  I would need to return 4 individual tokens right?

henrij22 replied:
I have added a return statement to the lower-level api for now, but not for the higher-level api like the element registration and loggers
https://github.com/ikarus-project/ikarus/blob/ea82beac8b9df1d3ea22feda1dfea5d03edf819b/ikarus/utils/listener/listener.hh#L63

henrij22 replied:
Here is a test for that
```cpp
  int i       = 0;
  auto callee = [&]() { i++; };
  auto token  = fe.subscribe(broadcaster, [&](UpdateMessages message) { callee(); });

  broadcaster.emitMessage(UpdateMessages::INCREMENT);
  t.check(i == 1) << testLocation();

  fe.unSubscribe(std::move(token));
  broadcaster.emitMessage(UpdateMessages::INCREMENT);
  t.check(i == 1) << testLocation();
```

rath3t replied:
I think rher are also some listeners, where subscription is optional and other where it is mandatory. Therefore, we may not habe to provide a mechanism to unsubscribe to complicated stuff if we have UB anywa if somebody unsubscribes something mandatory.

rath3t commented on docs/website/01_framework/controlRoutines.md:
````diff
@@ -9,13 +9,14 @@ status: new
 
 ```cpp
-auto lc = Ikarus::LoadControl(nonlinearSolver, numLoadSteps, {loadFactorStartValue, loadFactorEndValue});
+auto lc = Ikarus::LoadControl(nonlinearSolver, numLoadSteps, {loadFactorStartValue, loadFactorEndValue}, assembler);
````
This is outdated, isnt it?

rath3t commented on ikarus/controlroutines/controlroutinefactory.hh:
```diff
@@ -0,0 +1,62 @@
+  template <typename NLS, typename Assembler>
+  requires Concepts::FlatAssembler<typename std::remove_cvref_t<Assembler>::element_type>
+  auto create(NLS&& nonlinearSolver, Assembler&& assembler) const {
+    auto cr = createControlRoutine(std::move(config_), std::forward<NLS>(nonlinearSolver));
```
You move the config_ here. Maybe you can also pass it to the create function and make it static. The move seems a bit suprising here. Even though no user will stumble upon that.

henrij22 replied:
What exactly do you mean with "make it static" ? But yeah the `move` is a bit wierd here

henrij22 replied:
Ah so don't have config_ as member but pass it directly to `create` as well

henrij22 replied:
Fixed in the newest commit (newer then the Review Request)

rath3t reviewed (commented):
I think this is ready with the unsubscribe addition.

rath3t commented on tests/src/testelementsaslisteners.cpp:
```diff
@@ -0,0 +1,268 @@
+  using GridView  = typename Traits::GridView;
+  using Element   = typename Traits::Element;
+  using Pre       = DummySkillPre;
+  using NRState   = NonlinearSolverStateType<Requirement>;
```
NSState?

rath3t commented on ikarus/finiteelements/mixin.hh:
```diff
@@ -244,6 +247,36 @@ public:
+#else
+    if constexpr (requires { Sk::template subscribeToImpl<MT>(); }) {
+#endif
+      auto fTuple = Sk::template subscribeToImpl<MT>();
```
I would suggest that we not request a tuple of functions from the skill but we just forward, this, bc and the MT to the skill and the skill itself takes care to what it wants to subscribe. Then it can extract the non-linearsovler type from BC.

rath3t commented on tests/src/testelementsaslisteners.cpp:
```diff
@@ -0,0 +1,268 @@
+  template <typename MT>
+  auto subscribeToImpl() {
+    if constexpr (std::same_as<MT, NonLinearSolverMessages>) {
+      return std::make_tuple([&](NonLinearSolverMessages message, NRState& state) {
```
Here you can then extract the state from BC actually.

henrij22 replied:
Ah good idea!

rath3t commented on ikarus/solver/nonlinearsolver/nonlinearsolverbase.hh:
```diff
@@ -0,0 +1,24 @@
+namespace Ikarus {
+
+template <typename NLO, typename... Args>
+struct NonlinearSolverBase
```
I would remove the solver state class and just put that in here.

rath3t commented on ikarus/solver/nonlinearsolver/nonlinearsolverbase.hh:
```diff
@@ -0,0 +1,24 @@
+    : public Broadcasters<void(NonLinearSolverMessages), void(NonLinearSolverMessages, double),
+                          void(NonLinearSolverMessages, int),
+                          void(NonLinearSolverMessages, typename NonlinearSolverStateFactory<NLO>::type&), Args...>
+{
```
```suggestion
    {  using State = NonlinearSolverState<NLO>;
```

Then the state can check by itself what to do, with which NLO type

henrij22 replied:
I am not sure what you mean... above you said i should remove SolverState and now its in here?

rath3t commented on ikarus/controlroutines/controlroutinebase.hh:
```diff
@@ -0,0 +1,22 @@
+    : public Broadcasters<void(ControlMessages), void(ControlMessages, const std::string&),
+                          void(ControlMessages, int, const std::string&), void(ControlMessages, int, double)>
+
+{
```
Same here then, we also want a broadcaster with a controlroutine state. Then this base have to be templated though...

rath3t commented on ikarus/controlroutines/controlroutinebase.hh:
```diff
@@ -0,0 +1,32 @@
+struct ControlRoutineBase
+    : public Broadcasters<void(ControlMessages), void(ControlMessages, const std::string&),
+                          void(ControlMessages, int, const std::string&), void(ControlMessages, int, double),
+                          void(ControlMessages, ControlRoutineStateType<NLO>&), Args...>
```
this could be a const state ref or?

rath3t commented on ikarus/controlroutines/controlroutinestate.hh:
```diff
@@ -0,0 +1,60 @@
+// Forward declarations
+#ifndef DOXYGEN
+template <typename TypeListOne, typename TypeListTwo>
+class NonLinearOperator;
```
Nice

rath3t commented on tests/src/testelementsaslisteners.cpp:
```diff
@@ -0,0 +1,272 @@
+  auto calculateAtImpl(const Requirement& req, [[maybe_unused]] const Dune::FieldVector<double, Traits::mydim>& local,
+                       Dune::PriorityTag<0>) const {}
+
+  void updateState(NonLinearSolverMessages message, const Eigen::VectorXd& vec,
```
In the end we want to receive here in one way or the other the ferequirements. Therefore, we have to know the assembler somehow.

henrij22 replied:
Done

rath3t commented on ikarus/controlroutines/controlroutinefactory.hh:
```diff
@@ -0,0 +1,51 @@
+  static auto create(const CRConfig& config, NLS&& nonlinearSolver, Assembler&& assembler) {
+    auto cr = createControlRoutine(std::move(config), std::forward<NLS>(nonlinearSolver));
+
+    for (auto& fe : assembler->finiteElements()) {
```
So what if the element themself get a pointer to the current feRequirements?

henrij22 replied:
I think thats outdated now right?

rath3t replied:
yes

rath3t commented on ikarus/finiteelements/mixin.hh:
```diff
@@ -24,7 +28,7 @@ namespace Ikarus {
 template <typename PreFE, template <typename, typename> class... Skills>
-struct FEMixin : Skills<PreFE, typename PreFE::template FE<Skills...>>...
+struct FEMixin : public Listener, Skills<PreFE, typename PreFE::template FE<Skills...>>...
 {
```
Then also the calculateScalar,... method also get simpler and we don't have to pass the requirements all the time. Then the finite element knows the global current displacement vector by itself and the correction from the nonlinear solver state. But that a big change. What do you think @tarun-mitruka and @henrij22 ?

henrij22 replied:
I think this is out-of scope for this PR and maybe also v1.0 ?

rath3t commented on ikarus/controlroutines/controlroutinestate.hh:
```diff
@@ -0,0 +1,60 @@
+template <typename LoadParameter>
+struct ControlRoutineState
+{
+  LoadParameter parameter;
```
The Controlroutine state can just be templated with the controlroutine state and just contains a pointer to the const controlroutine + some other needed things maybe.

henrij22 replied:
Resolve as outdated

rath3t commented on ikarus/solver/nonlinearsolver/nonlinearsolverstate.hh:
```diff
@@ -0,0 +1,66 @@
+template <typename Correction, typename SolutionType = Correction>
+struct NonlinearSolverState
+{
+  Correction correction;
```
This should also  contain a pointer to the nonlinear solver. Then it can also access directly the nonlinear operator and what not.

henrij22 committed:
fix CHANGELOG.md

inital work

test run?

test runs

test runs

this wont work

broadcaster

will this work?

fix cmake

work on nonlinearsolverlogger

save

the observer is gone, long live the broadcaster (and the listener)

add dfdDD init in pathfollowing.inl
fix subsidiaryArgs

WIP on feature/updateStateObserver

rename stuff

register fes in arclength

fix tests

What is wrong here?

fix?

test + new element registration + tidy up

tidy up

idk

fix cmake

not a fix

factories for lc and pf

spelling

doxygen + fix

spelling

Code review part I

remove NoAssembler

test unsubscribeAll and using enum

changelog entry

comments PR Review

fix

introduce nonlinearsolverstate

this should work for all combinations ?

now for real

better factory for state

fix

add test for unsubscribe

fix controlroutinefactory

new mechanism to register broadcasters in skills

fix deduction of solverstate for tr

controlroutinestate

henrij22 committed:
Fixing merge conflict and test

henrij22 committed:
fixes

henrij22 committed:
fixis

henrij22 committed:
delete comment

rath3t commented on ikarus/controlroutines/pathfollowing.inl:
```diff
@@ -15,58 +15,67 @@
-  subsidiaryArgs.DD.setZero();
+  // info.totalIterations    = 0;
+  // subsidiaryArgs.stepSize = stepSize_;
+  // subsidiaryArgs.DD.resizeLike(req.glbalSolution());
```
Can you comment on that ?

henrij22 replied:
Ups that was from merging the above should be enough as subsidiaryArgs is now a member variable

rath3t commented on ikarus/controlroutines/pathfollowing.inl:
```diff
@@ -15,58 +15,67 @@
+  this->notify(SOLUTION_CHANGED);
+  this->notify(STEP_ENDED);
+
+  auto state = typename PathFollowing::State{.domain = req};
```
Would it be better to also include subsidiaryArgs_ in the state? Then these notifications look more the same? `this->notify(XXXXXX,state)`

henrij22 replied:
For now I have a generic `State` that works for all our Controlroutines. But yes I can add this

rath3t replied:
Yes thats also nice! But that state is anyway derivable at compile time as a template of the controlroutine. Orignuially that was the motivation of using listener and broadcaster, which could handle different templates due to the type erasure

henrij22 replied:
Done we have now a `PathfollowingState`

rath3t commented on ikarus/controlroutines/pathfollowing.inl:
```diff
@@ -15,58 +15,67 @@
+    state.loadStep = subsidiaryArgs_.currentStep;
+    state.stepSize = subsidiaryArgs_.stepSize;
+    this->notify(SOLUTION_CHANGED, state);
+    this->notify(SOLUTION_CHANGED);
```
Do we need both of them?

henrij22 replied:
Interesting, I think that was an artifact of resolving merge conflicts

henrij22 replied:
In general I think we can just always send (Message, state) what do do you think? Then for example the logger can make what it wants with it?

rath3t replied:
Yes. Thats would be the best i guess

rath3t commented on ikarus/utils/concepts.hh:
```diff
@@ -30,6 +30,9 @@ template <typename Derived>
 }
 
+template <typename T>
+concept EigenType = std::is_base_of_v<Eigen::MatrixBase<std::decay_t<T>>, std::decay_t<T>>;
```
This should be in our namespace?

rath3t commented on ikarus/utils/traits.hh:
```diff
@@ -147,6 +150,28 @@ struct isSharedPtr<std::shared_ptr<T>> : std::true_type
+};
+
+template <typename T, typename Deleter>
+struct isUniquePtr<std::unique_ptr<T, Deleter>> : std::true_type
```
Should'nt this specialization be enough?

rath3t commented on ikarus/utils/traits.hh:
```diff
@@ -8,13 +8,16 @@
 #include <tuple>
 #include <type_traits>
+// namespace std {
+// template <class T>
```
outdated comments

rath3t commented on ikarus/utils/listener/listener.hh:
```diff
@@ -0,0 +1,89 @@
+  requires(not Concepts::PointerOrSmartPointer<Broadcaster>)
+  auto subscribe(Broadcaster& broadcaster, F&& f) {
+    // if constexpr (requires { broadcaster.operator->(); })
+    //   t.push_back(broadcaster->template station<Signature>().registerListener(std::forward<F>(f)));
```
outdated

rath3t commented on ikarus/utils/functionhelper.hh:
```diff
@@ -107,4 +107,19 @@ auto referenceElementVertexPositions(FE& fe) {
+ * \param t
+ * \return decltype(auto)
+ */
+template <typename T>
```
I dont think this is the right place for that function but I also dont have a suggestion

henrij22 replied:
maybe we can put in traits.hh

rath3t commented on ikarus/utils/concepts.hh:
```diff
@@ -605,5 +608,45 @@ namespace Concepts {
+  struct MaybeDereference<T, true>
+  {
+    // We know T is pointer-like, so *std::declval<T&>() is well-formed.
+    using type = std::remove_reference_t<decltype(*std::declval<T&>())>;
```
Why not `std::remove_pointer`? https://en.cppreference.com/w/cpp/types/remove_pointer . Ah this only works for raw pointers

rath3t commented on ikarus/utils/derivativetraits.hh:
```diff
@@ -73,20 +77,23 @@ struct DerivativeTraitsFromCallables<Impl::Functions<DerivativeArgs...>, Arg>
   template <int I>
   using Range = std::tuple_element_t<I, Ranges>;
 
+  constexpr static int numberOfRanges = std::tuple_size_v<Ranges>;
```
This is not used? and in DifferentiableFunctions there is also a number indicating the number of derivatives

henrij22 replied:
Ah right, I didn't see it. I use this number in NLSState

rath3t commented on ikarus/utils/derivativetraits.hh:
```diff
@@ -73,20 +77,23 @@ struct DerivativeTraitsFromCallables<Impl::Functions<DerivativeArgs...>, Arg>
-    static constexpr int indexOfSignatureImpl = traits::Index<Signature, RawSignatures>::value + 1;
-    static constexpr int indexOfSignature     = traits::Index<Signature, RawSignatures>::value + 1;
+    // static constexpr int indexOfSignatureImpl = traits::Index<Signature, RawSignatures>::value + 1;
+    static constexpr int indexOfSignature = traits::Index<Signature, RawSignatures>::value + 1;
```
Outdated

rath3t reviewed (commented):
Thansk for this PR. see comments

henrij22 committed:
PR review I

rath3t commented on ikarus/controlroutines/controlroutinestate.hh:
```diff
@@ -0,0 +1,58 @@
+// Forward declarations
+#ifndef DOXYGEN
+template <typename TypeListOne, typename TypeListTwo>
+class NonLinearOperator;
```
This forward decleration is not used anymore or is it?

rath3t commented on ikarus/solver/nonlinearsolver/nonlinearsolverstate.hh:
```diff
@@ -0,0 +1,80 @@
+// Forward declarations
+#ifndef DOXYGEN
+template <typename TypeListOne, typename TypeListTwo>
+class NonLinearOperator;
```
Here too?

rath3t reviewed (approved):
See the unneed forward declerations, you can remove them and then you can merge

henrij22 committed:
Last changes from PR Review
